### PR TITLE
load data: use dist framework for single-node import

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #42930

Problem Summary:

### What is changed and how it works?
- add `GetEligibleInstances` to `TaskFlowHandle`
- single-node load data uses dist framework, since there're some tests cann't be run with dist framework, we use `__distributed` param and `EnableDistTask=false` to simulate.
  - and we change all realtikvtest of distributed load data use single node, since test env might run multiple test instances, and current distributed bind subtask to instance in round-robin way, might cause the test fail.
- use thread count of `load data` option for distribute load data
- fix some unstable case

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -414,33 +413,26 @@ func (d *dispatcher) processNormalFlow(gTask *proto.Task) (err error) {
...
-	pos = pos % len(serverNodes)
-	return serverNodes[pos].ID, nil
+// GetInstanceForSubtask gets an eligible instance.
+// we assign the subtask to the instance in a round-robin way.
+func GetInstanceForSubtask(serverNodes []*infosync.ServerInfo, taskIdx int) string {
+	pos := taskIdx % len(serverNodes)
+	return serverNodes[pos].ID
 }
 
 // GenerateSchedulerNodes generate a eligible TiDB nodes.
```
Comment (GMHDBJD) [inline disttask/framework/dispatcher/dispatcher.go:438]: Prefer move this function to TaskHandle, https://github.com/pingcap/tidb/blob/master/disttask/framework/dispatcher/dispatcher.go#L63, and pass the handle to bussiness dispatcher

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

File: disttask/framework/dispatcher/dispatcher.go (hunk 1 above)
Comment (D3Hunter) [reply disttask/framework/dispatcher/dispatcher.go:438]: let's change it when we support "subtask not bind to instance"

File: tests/realtikvtest/loaddatatest/BUILD.bazel [hunk 2]
```diff
@@ -2,7 +2,7 @@ load("@io_bazel_rules_go//go:def.bzl", "go_test")
 
 go_test(
     name = "loaddatatest_test",
-    timeout = "moderate",
+    timeout = "long",
```
Comment (D3Hunter) [inline tests/realtikvtest/loaddatatest/BUILD.bazel:5]: change to long to avoid timeout

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Comment by D3Hunter** [comment]
@lance6716 @zimulala

File: disttask/framework/proto/task.go [hunk 3]
```diff
@@ -53,11 +53,12 @@ const (
...
-	Key             string
-	Type            string
-	State           string
-	Step            int64
+	ID    int64
+	Key   string
+	Type  string
+	State string
+	Step  int64
+	// not used now.
```
Comment (zimulala) [inline disttask/framework/proto/task.go:61]:
```suggestion
	// DispatcherID is not used now.
```

File: executor/load_data.go [hunk 4]
```diff
@@ -290,13 +290,19 @@ func (e *LoadDataWorker) getJobImporter(ctx context.Context, job *asyncloaddata.
 		Progress: e.progress,
 	}
 
-	if variable.EnableDistTask.Load() && e.importPlan.Distributed {
-		return loaddata.NewDistImporter(param, e.importPlan, e.stmt)
+	if e.importPlan.Distributed {
+		// todo: right now some test cases will fail if we run single-node import using NewDistImporterCurrNode
```
Comment (zimulala) [inline executor/load_data.go:294]:
```suggestion
		// TODO: Right now some test cases will fail if we run single-node import using NewDistImporterCurrNode
```

File: executor/load_data.go [hunk 5]
```diff
@@ -290,13 +290,19 @@ func (e *LoadDataWorker) getJobImporter(ctx context.Context, job *asyncloaddata.
...
+		if variable.EnableDistTask.Load() {
+			return loaddata.NewDistImporter(param, e.importPlan, e.stmt)
+		}
+		return loaddata.NewDistImporterCurrNode(param, e.importPlan, e.stmt)
 	}
 
 	if e.controller.ImportMode == importer.LogicalImportMode {
 		return newLogicalJobImporter(param, e, r)
 	}
+	// todo: replace it with NewDistImporterCurrNode after we fix the test cases.
```
Comment (zimulala) [inline executor/load_data.go:305]:
```suggestion
	// TODO: Replace it with NewDistImporterCurrNode after we fix the test cases.
```

File: disttask/framework/dispatcher/dispatcher.go [hunk 6]
```diff
@@ -418,34 +417,26 @@ func (d *dispatcher) processNormalFlow(gTask *proto.Task) (err error) {
 	}
 
 	// Generate all available TiDB nodes for this global tasks.
-	serverNodes, err1 := GenerateSchedulerNodes(d.ctx)
+	serverNodes, err1 := handle.GetEligibleInstances(d.ctx, gTask)
 	if err1 != nil {
 		return err1
 	}
+	if len(serverNodes) == 0 {
+		return errors.New("no available TiDB node")
```
Comment (zimulala) [inline disttask/framework/dispatcher/dispatcher.go:425]: We do it in GenerateSchedulerNodes [line449](https://github.com/pingcap/tidb/pull/43880/files#diff-2a828989b540648a0bdef3d984d7946ea4fcb3c18718a4047b320c712ee1b855R449). Could we remove it?

File: disttask/framework/dispatcher/dispatcher.go [hunk 7]
```diff
@@ -418,34 +417,26 @@ func (d *dispatcher) processNormalFlow(gTask *proto.Task) (err error) {
...
 	return d.updateTask(gTask, gTask.State, subTasks, retrySQLTimes)
 }
 
-// GetEligibleInstance gets an eligible instance.
-func GetEligibleInstance(serverNodes []*infosync.ServerInfo, pos int) (string, error) {
-	if pos >= len(serverNodes) && pos < 0 {
-		errMsg := fmt.Sprintf("available TiDB nodes range is 0 to %d, but request position: %d", len(serverNodes)-1, pos)
-		return "", errors.New(errMsg)
-	}
-	if len(serverNodes) == 0 {
```
Comment (zimulala) [inline disttask/framework/dispatcher/dispatcher.go:443]: Why remove these check? My be once using this function，and len is 0.

Comment (D3Hunter) [reply disttask/framework/dispatcher/dispatcher.go:443]: previous `errMsg := fmt.Sprintf("available` is dead code, never reaches

below one move it to caller site `	if len(serverNodes) == 0 {
		return errors.New("no available TiDB node")`

File: disttask/framework/dispatcher/dispatcher.go (hunk 6 above)
Comment (D3Hunter) [reply disttask/framework/dispatcher/dispatcher.go:425]: this line is moved from `GetInstanceForSubtask` actually

File: disttask/framework/dispatcher/dispatcher.go (hunk 7 above)
Comment (zimulala) [reply disttask/framework/dispatcher/dispatcher.go:443]: It think this check put this function is more reasonable. After all, this function doesn't have to be called in this PR.

Comment (D3Hunter) [reply disttask/framework/dispatcher/dispatcher.go:443]: https://github.com/pingcap/tidb/pull/43880/commits/01bba1351dec9bb90ef49d6a1c940507b160cf0f

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**